### PR TITLE
refactor server to allow pluggable db and tracing

### PR DESCRIPTION
**Pull request by conradludgate**
Opens the way for third party servers to maintain and choose a database implementation themselves

### CODE REVIEW

**Comment by ellie** [comment]
Looks good at a first pass, will test and 👀 again though to make sure

File: atuin-server/src/handlers/history.rs [hunk 1]
```diff
@@ -140,8 +144,24 @@ pub async fn add<DB: Database>(
         })
         .collect();
 
-    let db = &state.0.database;
-    if let Err(e) = db.add_history(&history).await {
+    history.retain(|h| {
```
Comment (ellie) [inline atuin-server/src/handlers/history.rs:147]: This is changing the behaviour right? So if someone inserts something too large, their client will continue trying to sync

Comment (conradludgate) [reply atuin-server/src/handlers/history.rs:147]: Shouldn't be any change in functionality, I just move it from the database layer to the application layer https://github.com/ellie/atuin/blob/main/atuin-server/src/database.rs#L305-L318

Comment (ellie) [reply atuin-server/src/handlers/history.rs:147]: Ahhh of course! Makes sense, thanks

**Review by ellie — APPROVED** [review]
:shipit:

**Pull request merged**